File: algoritmos/IV.4 Parallel Block/src/c/IV.4-Parallel-Block.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <omp.h>
#include "../../../../utils/matrix_utils.c"

#define BLOCK_SIZE 2  // Block size

// Function to perform block matrix multiplication
void parallel_block_multiply(int64_t **A, int64_t **B, int64_t **C, int n, int block_size) {
    int i, j, k, ii, jj, kk;

    #pragma omp parallel for private(i, j, k, ii, jj, kk) shared(A, B, C)
    for (i = 0; i < n; i += block_size) {
        for (j = 0; j < n; j += block_size) {
            for (k = 0; k < n; k += block_size) {
                for (ii = i; ii < i + block_size && ii < n; ii++) {
                    for (jj = j; jj < j + block_size && jj < n; jj++) {
                        int64_t sum = 0; // Local sum for this block
                        for (kk = k; kk < k + block_size && kk < n; kk++) {
                            sum += A[ii][kk] * B[kk][jj];
                        }
                        C[ii][jj] += sum; // Directly accumulate
                    }
                }
            }
        }
    }
}

void print_matrix(int64_t **M, int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            printf("%ld ", M[i][j]);
        }
        printf("\n");
    }
}
// Main function
int main(int argc, char *argv[]) {
    int64_t **A, **B, **C;
    int n;
    const char *filename = argv[1];  // Path to the matrix file

    if (read_matrices_from_file(filename, &A, &B, &n) != 0) {
        fprintf(stderr, "Failed to read matrices from file.\n");
        return -1;
    }

    // Allocate memory for the result matrix C and initialize to zero
    C = (int64_t **)malloc(n * sizeof(int64_t *));
    for (int i = 0; i < n; i++) {
        C[i] = (int64_t *)malloc(n * sizeof(int64_t));
        memset(C[i], 0, n * sizeof(int64_t)); // Initialize to zero
    }

    // Perform parallel block matrix multiplication
    parallel_block_multiply(A, B, C, n, BLOCK_SIZE);

    // Print the result matrix
    printf("Result matrix C:\n");
    print_matrix(C, n);

    // Free allocated memory
    for (int i = 0; i < n; i++) {
        free(A[i]);
        free(B[i]);
        free(C[i]);
    }
    free(A);
    free(B);
    free(C);

    return 0;
}

File: algoritmos/IV.4 Parallel Block/src/python/IV.4-Parallel-Block.py
import numpy as np
import multiprocessing as mp

BLOCK_SIZE = 2  # Block size

def read_matrices_from_file(filename):
    # This function should read two matrices and their size from a file.
    # Replace this with actual file reading logic.
    # Example implementation with dummy matrices for testing:
    with open(filename, 'r') as f:
        # Read lines and strip any surrounding whitespace
        lines = [line.strip() for line in f.readlines() if line.strip()]

        # Determine the matrix size based on the number of rows up to the first blank line
        matrix_size = lines.index('') if '' in lines else len(lines) // 2
        n = matrix_size  # Assuming both matrices are square with the same size

        # Read the first matrix
        A = np.array( [[int(num) for num in line.split()] for line in lines[:n]], dtype=np.int64 )

        # Read the second matrix
        B = np.array( [[int(num) for num in line.split()] for line in lines[n:]], dtype=np.int64 )

    return A, B, n


def parallel_block_multiply_worker(args):
    """Performs block matrix multiplication for a specific block."""
    A, B, n, i, j, k, block_size = args
    C_local = np.zeros((block_size, block_size), dtype=np.int64)
    
    for ii in range(i, min(i + block_size, n)):
        for jj in range(j, min(j + block_size, n)):
            sum = 0
            for kk in range(k, min(k + block_size, n)):
                sum += A[ii, kk] * B[kk, jj]
            C_local[ii - i, jj - j] += sum
    return (i, j, C_local)

def parallel_block_multiply(A, B, n, block_size):
    C = np.zeros((n, n), dtype=np.int64)
    tasks = [
        (A, B, n, i, j, k, block_size)
        for i in range(0, n, block_size)
        for j in range(0, n, block_size)
        for k in range(0, n, block_size)
    ]

    with mp.Pool(mp.cpu_count()) as pool:
        results = pool.map(parallel_block_multiply_worker, tasks)
        for i, j, C_local in results:
            C[i:i + block_size, j:j + block_size] += C_local
    return C

def print_matrix(M):
    for row in M:
        print(" ".join(map(str, row)))

def main(filename):
    # Load matrices from file
    A, B, n = read_matrices_from_file(filename)

    # Perform parallel block matrix multiplication
    C = parallel_block_multiply(A, B, n, BLOCK_SIZE)

    # Print the result matrix
    print("Result matrix C:")
    print_matrix(C)

if __name__ == "__main__":
    import sys
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <input_file>")
        sys.exit(1)

    filename = sys.argv[1]
    main(filename)

